File: leetcode/day-18_shortest_distance_to_a_character.py
from typing import List


class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        n = len(s)
        distances = [0] * n

        prev = -n
        for i in range(n):
            if s[i] == c:
                prev = i
            distances[i] = i - prev

        for i in range(prev - 1, -1, -1):
            if s[i] == c:
                prev = i
            distances[i] = min(distances[i], prev - i)

        return distances

File: leetcode/day-18_shortest_distance_to_a_character.cpp
class Solution {
public:
    vector<int> shortestToChar(string s, char c) {
        int n = s.length();
        vector<int> distances(n);

        // First pass: Left to Right
        int prev = -n;
        for (int i = 0; i < n; ++i) {
            if (s[i] == c) {
                prev = i;
            }
            distances[i] = i - prev;
        }

        // Second pass: Right to Left
        for (int i = prev - 1; i >= 0; --i) {
            if (s[i] == c) {
                prev = i;
            }
            distances[i] = min(distances[i], prev - i);
        }

        return distances;
    }
};
